packages/firestore/src/core/pipeline-util.ts

```diff
@@ -242,41 +288,322 @@ export function toPipeline(query: Query, db: Firestore): Pipeline {
+      stage.accumulators as unknown as Map<string, Expr>
+    )})`;
+    if (stage.groups.size > 0) {
+      result = result + `grouping(${canonifyExprMap(stage.groups)})`;
+    }
+    return result;
+  }
+  if (stage instanceof Distinct) {
+    return `${stage.name}(${canonifyExprMap(stage.groups)})`;
+  }
+  if (stage instanceof CollectionSource) {
+    return `${stage.name}(${stage.collectionPath})`;
+  }
+  if (stage instanceof CollectionGroupSource) {
+    return `${stage.name}(${stage.collectionId})`;
+  }
+  if (stage instanceof DatabaseSource) {
+    return `${stage.name}()`;
+  }
+  if (stage instanceof DocumentsSource) {
+    return `${stage.name}(${stage.docPaths.sort()})`;
+  }
+  if (stage instanceof Where) {
+    return `${stage.name}(${canonifyExpr(stage.condition)})`;
+  }
+  if (stage instanceof FindNearest) {
+    const vector = stage._vectorValue.value.mapValue.fields![
+      VECTOR_MAP_VECTORS_KEY
+    ].arrayValue?.values?.map(value => value.doubleValue);
+    let result = `${stage.name}(${canonifyExpr(stage._field)},${
+      stage._distanceMeasure
+    },[${vector}]`;
+    if (!!stage._limit) {
+      result = result + `,${stage._limit}`;
+    }
+    if (!!stage._distanceField) {
+      result = result + `,${stage._distanceField}`;
+    }
+    return result + ')';
+  }
+  if (stage instanceof Limit) {
+    return `${stage.name}(${stage.limit})`;
+  }
+  if (stage instanceof Offset) {
+    return `${stage.name}(${stage.offset})`;
+  }
+  if (stage instanceof Select) {
+    return `${stage.name}(${canonifyExprMap(stage.projections)})`;
+  }
+  if (stage instanceof Sort) {
+    return `${stage.name}(${canonifySortOrderings(stage.orders)})`;
+  }
+
+  throw new Error(`Unrecognized stage ${stage.name}`);
+}
+
+function canonifyExprMap(map: Map<string, Expr>): string {
+  const sortedEntries = Array.from(map.entries()).sort();
+  return `${sortedEntries
+    .map(([key, val]) => `${key}=${canonifyExpr(val)}`)
+    .join(',')}`;
+}
+
+export function canonifyPipeline(p: CorePipeline): string;
+export function canonifyPipeline(p: CorePipeline): string {
+  return p.stages.map(s => canonifyStage(s)).join('|');
+}
+
+// TODO(pipeline): do a proper implementation for eq.
+export function pipelineEq(left: CorePipeline, right: CorePipeline): boolean {
+  return canonifyPipeline(left) === canonifyPipeline(right);
+}
+
+export type PipelineFlavor = 'exact' | 'augmented' | 'keyless';
+
+export type PipelineSourceType =
+  | 'collection'
+  | 'collection_group'
+  | 'database'
+  | 'documents';
+
+export function asCollectionPipelineAtPath(
+  pipeline: CorePipeline,
+  path: ResourcePath
+): CorePipeline {
+  const newStages = pipeline.stages.map(s => {
+    if (s instanceof CollectionGroupSource) {
+      return new CollectionSource(path.canonicalString());
+    }
+
+    return s;
+  });
+
+  return new CorePipeline(pipeline.serializer, newStages);
+}
+
+export type QueryOrPipeline = Query | CorePipeline;
```

Comment:
Because QueryOrPipeline is a union type that is used throughout the implementation, to support porting to other platforms, we should change this to a common interface between the two types.